### PR TITLE
fix variables function

### PR DESCRIPTION
**Pull request by arcondello**
Add `fix_variables` function which in polynomial time can determine some variables in a bqm to fix.

This change also adds C++ code to dimod, necessitating some CI infrastructure changes.

This PR does not update the deploy in CI.

### CODE REVIEW

File: dimod/roof_duality/fix_variables.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+from dimod.roof_duality._fix_variables import fix_variables_wrapper
+from dimod.vartypes import Vartype
+
+
+def fix_variables(bqm, sampling_mode=True):
+    """Determine assignments for some bqm variables.
```
Comment (JoelPasvolsky) [inline dimod/roof_duality/fix_variables.py]: BQM, and elsewhere

File: dimod/roof_duality/fix_variables.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    fix_variables uses maximum flow in the implication network to correctly fix variables (that is,
+    one can find an assignment for the other variables that attains the optimal value). The
+    variables that roof duality fixes will take the same values in all optimal solutions.
+
+    Args:
+        bqm (:obj:`.BinaryQuadraticModel`)
+            A binary quadratic model.
+
+        sampling_mode (bool, optional, default=True):
```
Comment (JoelPasvolsky) [inline dimod/roof_duality/fix_variables.py]: sampling_mode description is unclear. What is sampling mode (i.e., what does it do for true vs false)?

File: dimod/roof_duality/fix_variables.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        dict: Variable assignments for some bqm variables.
+
+    Examples:
+
+        >>> bqm = dimod.BinaryQuadraticModel.empty(dimod.SPIN)
+        >>> bqm.add_variable('a', 1.0)
+        >>> dimod.fix_variables(bqm)
+        {'a': -1}
+
+        There are two ground states, so no variables fixed.
```
Comment (JoelPasvolsky) [inline dimod/roof_duality/fix_variables.py]: In the next example, there

(Otherwise looks like it's for the previous example.)

**Comment by arcondello** [comment]
Just realised I still need to add the Apache headers, will do so before merging

File: dimod/roof_duality/fix_variables.py (hunk 2 above)
Comment (randomir) [reply dimod/roof_duality/fix_variables.py]: Agree, it's not clear you get multiple samples for the same ground state energy when `sampling_mode=True`.

**Comment by arcondello** [comment]
closes #216

**Pull request merged**